pkg/apis/serving/v1beta1/revision_types.go

```diff
@@ -0,0 +1,136 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"github.com/knative/pkg/apis"
+	duckv1beta1 "github.com/knative/pkg/apis/duck/v1beta1"
+	"github.com/knative/pkg/kmeta"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Revision is an immutable snapshot of code and configuration.  A revision
+// references a container image. Revisions are created by updates to a
+// Configuration.
+//
+// See also: https://github.com/knative/serving/blob/master/docs/spec/overview.md#revision
+type Revision struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +optional
+	Spec RevisionSpec `json:"spec,omitempty"`
+
+	// +optional
+	Status RevisionStatus `json:"status,omitempty"`
+}
+
+// Verify that Revision adheres to the appropriate interfaces.
+var (
+	// Check that Revision can be validated, can be defaulted, and has immutable fields.
+	_ apis.Validatable = (*Revision)(nil)
+	_ apis.Defaultable = (*Revision)(nil)
+	_ apis.Immutable   = (*Revision)(nil)
+
+	// Check that we can create OwnerReferences to a Revision.
+	_ kmeta.OwnerRefable = (*Revision)(nil)
+)
+
+// RevisionTemplateSpec describes the data a revision should have when created from a template.
+// Based on: https://github.com/kubernetes/api/blob/e771f807/core/v1/types.go#L3179-L3190
+type RevisionTemplateSpec struct {
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +optional
+	Spec RevisionSpec `json:"spec,omitempty"`
+}
+
+// RevisionContainerConcurrencyType is an integer expressing a number of
+// in-flight (concurrent) requests.
+type RevisionContainerConcurrencyType int64
+
+const (
+	// The maximum configurable container concurrency.
+	RevisionContainerConcurrencyMax RevisionContainerConcurrencyType = 1000
+)
+
+// RevisionSpec holds the desired state of the Revision (from the client).
+type RevisionSpec struct {
+	// RevisionSpec inlines PodSpec and restricts the fields it allows
+	// via webhook.
+	corev1.PodSpec `json:",inline"`
+
+	// ContainerConcurrency specifies the maximum allowed
+	// in-flight (concurrent) requests per container of the Revision.
+	// Defaults to `0` which means unlimited concurrency.
+	// This field replaces ConcurrencyModel. A value of `1`
+	// is equivalent to `Single` and `0` is equivalent to `Multi`.
+	// +optional
+	ContainerConcurrency RevisionContainerConcurrencyType `json:"containerConcurrency,omitempty"`
+
+	// TimeoutSeconds holds the max duration the instance is allowed for responding to a request.
+	// +optional
+	TimeoutSeconds int64 `json:"timeoutSeconds,omitempty"`
```

Comment:
We should make this a pointer since the default value should not be 0. That way we can detect between a user set 0 and an 'unset'.

https://github.com/knative/serving/issues/3342

Reply:
🤔 What would a user set zero mean?

Reply:
Changed